README.md
# zellij-runner
<!-- cargo-sync-readme start -->

Below is the original README.md from [alex35mil](https://github.com/alex35mil/dotfiles/tree/master/user/bin/zellij/runner)
Thanks yo!

----------------------

Ad-hoc replacement of Zellij session switcher (which doesn't exist yet).

<img width="1329" alt="screen" src="https://user-images.githubusercontent.com/4244251/221364651-2011f53f-eaa4-445d-959e-86584ed1ad38.png">

## Installation

NOTE: if you follow the cargo install instructions below you won't be getting this mod


```sh
cargo install zellij-runner
```

## Usage
```sh
# run switcher in interactive mode
zellij-runner

# create/connect to specified session
zellij-runner my-session

# create session with specified layout
zellij-runner my-session my-layout
```

To exit the runner, hit `Esc` on session selector screen, or `Ctrl-C` on any screen.

## Configuration
### Layouts
The runner can include layout selector when creating a new session.
To activate it, set an environment variable with a path to the layouts folder:

```sh
ZELLIJ_RUNNER_LAYOUTS_DIR=.config/zellij/layouts
```

### Banner
To show a banner, provide a path to the directory with ASCII art.

```sh
ZELLIJ_RUNNER_BANNERS_DIR=.config/zellij/banners
```

Each file with ASCII art must have `.banner` extension.

Runner would pick a random banner each time you switch sessions.

### Paths autocompletion
To optimize autocompletion when switching working dir, set the following environment variables:

```sh
# directory with the projects, relative to the HOME dir
ZELLIJ_RUNNER_ROOT_DIR=Projects

# switcher already respects gitignore, but it's still useful in case there's no git
ZELLIJ_RUNNER_IGNORE_DIRS=node_modules,target

# traverse dirs 3 level max from ZELLIJ_RUNNER_ROOT_DIR
ZELLIJ_RUNNER_MAX_DIRS_DEPTH=3
```

<!-- cargo-sync-readme end -->


CHANGELOG.md
# Changelog

## 0.3.0
- Allow specifying layout but not session - layout ignored if attaching to existing session
- Use clap to parse args

## 0.3.0
- When changing a current directory for the new session, lowercase both dir and input during a filtering of the directory list.

## 0.2.0
- On session selector screen, `Ctrl-N` starts creating a new session.
- When creating a new session, `Esc` cancels it and takes to the session selector screen.
- `Ctrl-C` exits the runner from any screen.

## 0.1.0
Initial release.


src/action.rs
use std::{io, process};

use crate::{dir::Dir, log, zellij};

pub(crate) enum Action {
    AttachToSession(String),
    CreateNewSession {
        session: Option<String>,
        layout: Option<String>,
        dir: Option<Dir>,
    },
    Exit(Result<(), io::Error>),
}

impl Action {
    pub(crate) fn exec(self) {
        let action = self;

        let status = match action {
            Action::CreateNewSession {
                session,
                layout,
                dir: wd,
            } => zellij::create(&session, &layout, &wd),
            Action::AttachToSession(session) => zellij::attach(&session),
            Action::Exit(Ok(())) => process::exit(0),
            Action::Exit(Err(error)) => Self::exit_with_error(error),
        };

        match status {
            Ok(status) => {
                if !status.success() {
                    process::exit(status.code().unwrap_or(1));
                }
            }
            Err(err) => Self::exit_with_error(err),
        }
    }

    fn exit_with_error(error: io::Error) -> ! {
        log::error(error);
        process::exit(1);
    }
}


src/runner.rs
use crate::{action::Action, ui, zellij, Args};

pub(crate) fn init(args: Args) {
    let action = match zellij::list_sessions() {
        Err(error) => Action::Exit(Err(error)),
        Ok(sessions) => match (args.session, args.layout, sessions.as_slice()) {
            (None, layout, &[]) => {
                // TODO: add optional layout specification
                ui::new_session_prompt(sessions, layout, args.chdir)
            }
            (session, layout, &[]) => Action::CreateNewSession {
                session,
                layout,
                dir: None,
            },
            // TODO: make "prompt" arg ?
            (Some(session), layout, _) => {
                if sessions.contains(&session) {
                    // just ignore layout if attaching to session -- for now
                    // maybe there's a way to attach to a session then apply a new layout???
                    Action::AttachToSession(session)
                } else {
                    Action::CreateNewSession {
                        session: Some(session),
                        layout,
                        dir: None,
                    }
                }
            }
            (None, layout, _) => ui::action_selector(sessions, layout),
        },
    };

    action.exec()
}

pub(crate) fn switch(args: Args) {
    // Note: probably not the best way to do this, but it's fine
    let action = match zellij::list_sessions() {
        Err(error) => Action::Exit(Err(error)),
        Ok(sessions) => match sessions.as_slice() {
            // TODO: keep layout somewhere global and use?? prolly not
            &[] => ui::new_session_prompt(sessions, None, args.chdir),
            _ => ui::action_selector(sessions, None),
        },
    };

    action.exec()
}


src/zellij.rs
use std::{
    fs, io,
    process::{Command, ExitStatus},
};

use crate::dir::Dir;

const BIN: &str = "zellij";

pub(crate) fn list_sessions() -> io::Result<Vec<String>> {
    let output = Command::new(BIN).arg("list-sessions").output()?;

    if output.status.success() {
        let stdout = String::from_utf8(output.stdout).unwrap();
        let lines: Vec<String> = stdout
            .split('\n')
            .into_iter()
            .filter_map(|s| match s {
                "" => None,
                s => Some(s.to_owned()),
            })
            .collect();
        Ok(lines)
    } else {
        let exit_code = match output.status.code() {
            Some(code) => code.to_string(),
            None => "-".to_string(),
        };

        let stderr = String::from_utf8(output.stderr);

        match stderr {
            Ok(err) => {
                if err.contains("No") && err.contains("sessions found") {
                    Ok(vec![])
                } else {
                    Err(io::Error::new(
                        io::ErrorKind::Other,
                        format!(
                            "Failed to get Zellij sessions. Exit code: {}. {}",
                            exit_code, err
                        ),
                    ))
                }
            }
            Err(_) => Err(io::Error::new(
                io::ErrorKind::Other,
                format!("Failed to get Zellij sessions. Exit code: {}", exit_code),
            )),
        }
    }
}

pub(crate) fn list_layouts(dir: &Dir) -> io::Result<Vec<String>> {
    let layouts = match fs::read_dir(dir) {
        Ok(entries) => {
            let size = entries.size_hint().1;
            let mut layouts = match size {
                Some(size) => Vec::with_capacity(size),
                None => vec![],
            };
            for entry in entries {
                match entry {
                    Ok(entry) => {
                        let path = entry.path();
                        if let Some(ext) = path.extension() {
                            if ext.to_string_lossy() == "kdl" {
                                let layout =
                                    path.file_stem().unwrap().to_string_lossy().to_string();
                                layouts.push(layout);
                            }
                        }
                    }
                    Err(err) => return Err(err),
                }
            }
            layouts
        }
        Err(err) => return Err(err),
    };

    Ok(layouts)
}

pub(crate) fn create(
    session: &Option<String>,
    layout: &Option<String>,
    dir: &Option<Dir>,
) -> Result<ExitStatus, io::Error> {
    let mut args = vec![];

    if let Some(session) = session {
        args.extend_from_slice(&["--session", session]);
    }

    if let Some(layout) = layout {
        args.extend_from_slice(&["--layout", layout]);
    }

    let mut cmd = Command::new(BIN);

    if let Some(dir) = dir {
        cmd.current_dir(dir);
    }

    cmd.args(&args).status()
}

pub(crate) fn attach(session: &str) -> Result<ExitStatus, io::Error> {
    Command::new(BIN).args(["attach", session]).status()
}


src/main.rs
//! Ad-hoc replacement of Zellij session switcher (which doesn't exist yet).
//!
//! <img width="1329" alt="screen" src="https://user-images.githubusercontent.com/4244251/221364651-2011f53f-eaa4-445d-959e-86584ed1ad38.png">
//!
//! ## Installation
//! ```sh
//! cargo install zellij-runner
//! ```
//!
//! ## Usage
//! ```sh
//! # run switcher in interactive mode
//! zellij-runner
//!
//! # create/connect to specified session
//! zellij-runner my-session
//!
//! # create session with specified layout
//! zellij-runner my-session my-layout
//! ```
//!
//! To exit the runner, hit `Esc` at any point.
//!
//! ## Configuration
//! ### Layouts
//! The runner can include layout selector when creating a new session.
//! To activate it, set an environment variable with a path to the layouts folder:
//!
//! ```sh
//! ZELLIJ_RUNNER_LAYOUTS_DIR=.config/zellij/layouts
//! ```
//!
//! ### Banner
//! To show a banner, provide a path to the directory with ASCII art.
//!
//! ```sh
//! ZELLIJ_RUNNER_BANNERS_DIR=.config/zellij/banners
//! ```
//!
//! Each file with ASCII art must have `.banner` extension.
//!
//! Runner would pick a random banner each time you switch sessions.
//!
//! ### Paths autocompletion
//! To optimize autocompletion when switching working dir, set the following environment variables:
//!
//! ```sh
//! # directory with the projects, relative to the HOME dir
//! ZELLIJ_RUNNER_ROOT_DIR=Projects
//!
//! # switcher already respects gitignore, but it's still useful in case there's no git
//! ZELLIJ_RUNNER_IGNORE_DIRS=node_modules,target
//!
//! # traverse dirs 3 level max from ZELLIJ_RUNNER_ROOT_DIR
//! ZELLIJ_RUNNER_MAX_DIRS_DEPTH=3
//! ```

use clap::Parser;

mod action;
mod dir;
mod log;
mod options;
mod runner;
mod ui;
mod zellij;

#[derive(Parser, Debug, Clone)]
#[clap(name = "zellij-runner")]
struct Args {
    #[arg(short, long)]
    session: Option<String>,
    #[arg(short, long)]
    layout: Option<String>,
    #[arg(long, action)]
    chdir: bool,
}

fn main() {
    let args = Args::parse();

    runner::init(args.clone());

    loop {
        runner::switch(args.clone());
    }
}
